domain/application/service/application.go

```diff
@@ -480,8 +534,51 @@ func (s *ApplicationService) insertCAASUnit(
 // If the application still has units, as error satisfying [applicationerrors.ApplicationHasUnits]
 // is returned.
 func (s *ApplicationService) DeleteApplication(ctx context.Context, name string) error {
-	err := s.st.DeleteApplication(ctx, name)
-	return errors.Annotatef(err, "deleting application %q", name)
+	var cleanups []func(context.Context)
+	err := s.st.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+		var err error
+		cleanups, err = s.deleteApplication(ctx, name)
+		return errors.Trace(err)
+	})
+	if err != nil {
+		return errors.Annotatef(err, "deleting application %q", name)
+	}
+	for _, cleanup := range cleanups {
+		cleanup(ctx)
+	}
+	return nil
+}
+
+func (s *ApplicationService) deleteApplication(ctx domain.AtomicContext, name string) ([]func(context.Context), error) {
+	// Get app owned secrets.
+	uris, err := s.st.GetSecretsForOwners(ctx, application.ApplicationSecretOwners{name}, application.NilUnitOwners)
+	if err != nil {
+		return nil, errors.Annotatef(err, "getting application owned secrets for %q", name)
+	}
+	// Delete app owned secrets.
+	var cleanups []func(context.Context)
+	for _, uri := range uris {
+		s.logger.Debugf("deleting application %q secret: %s", name, uri.ID)
+		cleanup, err := s.secretService.InternalDeleteSecret(ctx, uri, secretservice.DeleteSecretParams{
+			// TODO(units) - access is expected to be a unit
+			// It is passed to CleanupSecrets() on k8s and is used
+			// to name the service account..
+			Accessor: secretservice.SecretAccessor{
+				Kind: secretservice.UnitAccessor,
+				ID:   name + "/0",
```

Comment:
Should it be the leader unit?

Reply:
no, in 3.6 it is the unit which makes the api call. it's only used to name the name for the k8s service account.
but it currently must be a unit.
this is a temp fix until we can decide the best way to adapt it to 4.0